Upload task needs to be able to handle dev releases

```
2015-11-13 04:22:20.055: Build Summary
2015-11-13 04:22:20.055:              Project: cschcs-pyb-plugin
2015-11-13 04:22:20.055:              Version: 0.0.1.dev20151113042219
2015-11-13 04:22:20.055:       Base directory: /home/ec2-user/workspace/HCS_-_PyBuilder_Plugin_-_(master)_-_(Autogen)
2015-11-13 04:22:20.055:         Environments: 
2015-11-13 04:22:20.055:                Tasks: prepare [297 ms] analyze [143 ms] compile_sources [0 ms] run_unit_tests [21 ms] package [1 ms] run_integration_tests [8 ms] verify [123 ms] publish [303 ms]
2015-11-13 04:22:20.055: Build finished at 2015-11-13 04:22:20
2015-11-13 04:22:20.055: Build took 0 seconds (907 ms)
2015-11-13 04:22:20.062: Deactivating PyEnv installer...
2015-11-13 04:22:20.062: Deactivating NodeEnv...
2015-11-13 04:22:20.063: Deactivating VirtualEnv...
2015-11-13 04:22:20.266: [HCS_-_PyBuilder_Plugin_-_(master)_-_(Autogen)] $ /bin/bash +eEu /tmp/hudson4840723731945904239.sh
2015-11-13 04:22:20.268: Activating VirtualEnv...
2015-11-13 04:22:20.272: Activating NodeEnv...
2015-11-13 04:22:20.273: Activating PyEnv installer...
2015-11-13 04:22:20.324: PyBuilder version 0.11.3.dev20151112204412
2015-11-13 04:22:20.324: Build started at 2015-11-13 04:22:20
2015-11-13 04:22:20.324: ------------------------------------------------------------
2015-11-13 04:22:20.324: [INFO]  Building cschcs-pyb-plugin version 0.0.1.dev20151113042220
2015-11-13 04:22:20.324: [INFO]  Executing build in /home/ec2-user/workspace/HCS_-_PyBuilder_Plugin_-_(master)_-_(Autogen)
2015-11-13 04:22:20.324: [INFO]  Going to execute task upload
2015-11-13 04:22:20.324: [INFO]  Uploading project cschcs-pyb-plugin-0.0.1.dev20151113042220 to repository 'koji-pypi'
2015-11-13 04:22:20.324: ------------------------------------------------------------
2015-11-13 04:22:20.324: BUILD FAILED - [Errno 2] No such file or directory: '/home/ec2-user/workspace/HCS_-_PyBuilder_Plugin_-_(master)_-_(Autogen)/target/dist/cschcs-pyb-plugin-0.0.1.dev20151113042220'
2015-11-13 04:22:20.324: ------------------------------------------------------------
2015-11-13 04:22:20.324: Build finished at 2015-11-13 04:22:20
2015-11-13 04:22:20.324: Build took 0 seconds (11 ms)
```

The cause is that when `upload` is run in a separate task it generates its own timestamp and can't find the **latest** built snapshot.

What should happen is that upload should upload the latest artifact if the version of the project ends with '.dev'.

--- a/src/main/python/pybuilder/__init__.py
+++ b/src/main/python/pybuilder/__init__.py
@@ -18,7 +18,7 @@
 
 from pybuilder.errors import BuildFailedException
 
-__version__ = "${version}"
+__version__ = "${dist_version}"
 
 
 def bootstrap():

--- a/src/main/python/pybuilder/cli.py
+++ b/src/main/python/pybuilder/cli.py
@@ -37,7 +37,7 @@ from pybuilder.scaffolding import start_project
 from pybuilder.terminal import (BOLD, BROWN, RED, GREEN, bold, styled_text,
                                 fg, italic, print_text, print_text_line,
                                 print_error, print_error_line, draw_line)
-from pybuilder.utils import format_timestamp
+from pybuilder.utils import format_timestamp, get_dist_version_string
 
 PROPERTY_OVERRIDE_PATTERN = re.compile(r'^[a-zA-Z0-9_]+=.*')
 
@@ -212,7 +212,7 @@ def init_logger(options):
 def print_build_summary(options, summary):
     print_text_line("Build Summary")
     print_text_line("%20s: %s" % ("Project", summary.project.name))
-    print_text_line("%20s: %s" % ("Version", summary.project.version))
+    print_text_line("%20s: %s%s" % ("Version", summary.project.version, get_dist_version_string(summary.project)))
     print_text_line("%20s: %s" % ("Base directory", summary.project.basedir))
     print_text_line("%20s: %s" %
                     ("Environments", ", ".join(options.environments)))

--- a/src/main/python/pybuilder/core.py
+++ b/src/main/python/pybuilder/core.py
@@ -273,7 +273,9 @@ class Project(object):
 
     def __init__(self, basedir, version="1.0.dev0", name=None):
         self.name = name
-        self._version = version
+        self._version = None
+        self._dist_version = None
+        self.version = version
         self.basedir = basedir
         if not self.name:
             self.name = os.path.basename(basedir)
@@ -306,10 +308,14 @@ class Project(object):
 
     @version.setter
     def version(self, value):
-        assert isinstance(value, str)
+        self._version = value
         if value.endswith('.dev'):
             value += datetime.utcnow().strftime("%Y%m%d%H%M%S")
-        self._version = value
+        self._dist_version = value
+
+    @property
+    def dist_version(self):
+        return self._dist_version
 
     def validate(self):
         """

--- a/src/main/python/pybuilder/plugins/python/distutils_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/distutils_plugin.py
@@ -37,7 +37,7 @@ from pybuilder.core import (after,
                             RequirementsFile,
                             Dependency)
 from pybuilder.errors import BuildFailedException
-from pybuilder.utils import as_list, is_string, is_notstr_iterable
+from pybuilder.utils import as_list, is_string, is_notstr_iterable, get_dist_version_string
 
 from .setuptools_plugin_helper import build_dependency_version_string
 from textwrap import dedent
@@ -128,7 +128,7 @@ def render_setup_script(project):
     template_values = {
         "module": "setuptools" if project.get_property("distutils_use_setuptools") else "distutils.core",
         "name": project.name,
-        "version": project.version,
+        "version": project.dist_version,
         "summary": default(project.summary),
         "description": default(project.description),
         "author": author,
@@ -213,8 +213,9 @@ def upload(project, logger):
     upload_cmd_line.append("upload")
     upload_cmd_line.extend(repository_args)
 
-    logger.info("Uploading project %s-%s%s", project.name, project.version,
-                (" to repository '%s'" % repository) if repository else "")
+    logger.info("Uploading project %s-%s%s%s", project.name, project.version,
+                (" to repository '%s'" % repository) if repository else "",
+                get_dist_version_string(project, " as version %s"))
     execute_distutils(project, logger, [upload_cmd_line])
 
 

--- a/src/main/python/pybuilder/reactor.py
+++ b/src/main/python/pybuilder/reactor.py
@@ -36,7 +36,7 @@ from pybuilder.pluginloader import (BuiltinPluginLoader,
                                     DispatchingPluginLoader,
                                     ThirdPartyPluginLoader,
                                     DownloadingPluginLoader)
-from pybuilder.utils import as_list
+from pybuilder.utils import as_list, get_dist_version_string
 from pybuilder.execution import Action, Initializer, Task
 
 
@@ -147,7 +147,7 @@ class Reactor(object):
             [task.name for task in execution_plan]))
 
         self.logger.info(
-            "Building %s version %s", self.project.name, self.project.version)
+            "Building %s version %s%s", self.project.name, self.project.version, get_dist_version_string(self.project))
         self.logger.info("Executing build in %s", self.project.basedir)
 
         if len(tasks) == 1:

--- a/src/main/python/pybuilder/utils.py
+++ b/src/main/python/pybuilder/utils.py
@@ -596,3 +596,7 @@ else:
 def is_notstr_iterable(obj):
     """Checks if obj is iterable, but not a string"""
     return not isinstance(obj, basestring) and isinstance(obj, collections.Iterable)
+
+
+def get_dist_version_string(project, format=" (%s)"):
+    return format % project.dist_version if project.version != project.dist_version else ""

